### PR TITLE
Implementing advanced data tables

### PR DESCRIPTION
**Pull request by carlo**

### CODE REVIEW

**Comment by PeterSiegl** [comment]
Good job! Just two suggestions:
- We should truncate cells with long text values (maximum width + slds-truncate or via className)
- It would be cool to include responsiveness. we could either add auto scrolling on smaller devices or add stacking in a later version.

**Comment by fspoettel** [comment]
@PeterSiegl We should stick as close as possible to the Salesforce examples for the tables, otherwise a new version will bite us hard here. That's why I think that auto-scrolling is something we should wrap in the product. We could add it as a HOC. The stacked table differs a fair bit in markup, let's add it later as you suggested.

**Comment by extronics** [comment]
I did a clean rebase onto master. Feel free to give it another look :)

File: src/components/DataTable/index.js [hunk 1]
```diff
@@ -1,9 +1,4 @@
-import Cell from './Cell';
-import Row from './Row';
-import Table from './Table';
-
-export {
-  Cell,
-  Row,
-  Table
-};
+export { default as DataTableColumn } from './DataTableColumn';
```
Comment (fspoettel) [inline src/components/DataTable/index.js]: Please use `import` and `export` here. This syntax leads to the PropTypes and PropDescriptions not being picked up in Storybook.

File: stories/DataTable.js [hunk 2]
```diff
@@ -1,68 +1,156 @@
...
+        title="Column 1"
+      />
+      <DataTableColumn
+        dataKey="col3"
+        sortable
+        title="Column 3"
+      />
+      <DataTableActionColumn
+        dataKey="stuff"
+        cellRenderer={({ dataKey, rowIndex, defaultProps }) => (
```
Comment (fspoettel) [inline stories/DataTable.js]: There is a `defaultActionHeadRenderer`. Shouldn't this use it? Looks a bit broken:

screen shot 2017-12-08 at 11 02 56

File: stories/DataTable.js [hunk 3]
```diff
@@ -1,68 +1,156 @@
...
+          <td {...defaultProps}>
+            <span>
+              <strong>{dataKey}:</strong> {cellData}
+            </span>
+          </td>
+        )}
+      />
+    </DataTable>
+  ))
+  .add('Sortable table', () => (
```
Comment (fspoettel) [inline stories/DataTable.js]: Please add a couple more rows to this example, it is a bit confusing

File: src/components/DataTable/defaultHeadRenderer.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  const onClick = sortable ? () => onSort(dataKey) : null;
+  const classNames = cx('slds-text-title--caps', { 'slds-is-sorted': sortBy === dataKey });
+
+  return (
+    <Cell
+      className={classNames}
+      key={dataKey}
+      scope="col"
+      sortable={sortable}
+      sortDirection={sortDirection}
```
Comment (fspoettel) [inline src/components/DataTable/defaultHeadRenderer.js]: A `sortDirection` is passed down, yet it is not used in `Cell.js`.

File: src/components/DataTable/defaultSelectAllRenderer.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  return (
+    <th
+      className="slds-text-align--right"
+      key={dataKey}
+      scope="col"
+      style={{ width: '3.25rem' }}
+    >
+      <div>
+        <span className="slds-checkbox">
+          <input
```
Comment (fspoettel) [inline src/components/DataTable/defaultSelectAllRenderer.js]: This replicates `CheckbowRaw`, is there a reason it is not used here?

File: src/components/DataTable/defaultSelectRenderer.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  return (
+    <td
+      className="slds-text-align--right"
+      key={dataKey}
+      role="gridcell"
+      style={{ width: '3.25rem' }}
+    >
+      <span className="slds-checkbox">
+        <input
```
Comment (fspoettel) [inline src/components/DataTable/defaultSelectRenderer.js]: see `CheckboxRaw` comment

File: src/components/DataTable/DataTable.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+      onSort({ dataKey: nextSortBy, sortDirection: nextSortDirection });
+      this.setState(nextState);
+    } else {
+      this.setState({
+        ...nextState,
+        data: [...data].sort(
+          (a, b) => (
+            nextSortDirection === 'asc'
+              ? String(a[nextSortBy]).localeCompare(b[nextSortBy])
+              : String(b[nextSortBy]).localeCompare(a[nextSortBy])
```
Comment (fspoettel) [inline src/components/DataTable/DataTable.js]: 👍

File: stories/DataTable.js [hunk 8]
```diff
@@ -1,68 +1,156 @@
...
+        title="Column 1"
+      />
+      <DataTableColumn
+        dataKey="col3"
+        sortable
+        title="Column 3"
+      />
+    </DataTable>
+  ))
+  .add('Selectable table', () => (
```
Comment (fspoettel) [inline stories/DataTable.js]: We should clear up how selection works, there is no way to infer that from the docs. Maybe we could use the storybooks-`info` HoC for that.

**Comment by fspoettel** [comment]
closes #97

File: src/components/DataTable/defaultSelectAllRenderer.js (hunk 5 above)
Comment (extronics) [reply src/components/DataTable/defaultSelectAllRenderer.js]: No, will change 👍

File: src/components/DataTable/defaultSelectRenderer.js (hunk 6 above)
Comment (extronics) [reply src/components/DataTable/defaultSelectRenderer.js]: 👍

File: src/components/DataTable/index.js (hunk 1 above)
Comment (extronics) [reply src/components/DataTable/index.js]: 👍

File: stories/DataTable.js (hunk 3 above)
Comment (extronics) [reply stories/DataTable.js]: 👍

File: stories/DataTable.js (hunk 2 above)
Comment (extronics) [reply stories/DataTable.js]: It does (it's the default in `DataTableActionColumn.defaultProps`). What browser is that? Looks normal for me in FF/Chrome.

File: stories/Table.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import { Table, Row, Cell } from '../src';
+
+const stories = storiesOf('Table', module);
+
+stories
+  .add('Basic example', () => (
+    <Table
+      flavor={array('Flavor', ['bordered', 'striped']) || undefined}
+      variation={array('Variation', [])}
+
```
[the remark this answers is not in the document]
Comment (extronics) [reply stories/Table.js]: 👍

File: stories/DataTable.js (hunk 8 above)
Comment (extronics) [reply stories/DataTable.js]: 👍

File: stories/DataTable.js (hunk 2 above)
Comment (fspoettel) [reply stories/DataTable.js]: FF, it renders:

``` html
<div class="slds-th__action">
<span class="slds-assistive-text">Actions</span>
</div>
```

which explains the absence of visible title

**Pull request merged**